src/datastructure/RailwayNetwork.cpp

```diff
@@ -653,3 +657,749 @@ cda_rail::Network::Network(const std::filesystem::path &p) {
+                        ret_val.back().emplace_back(edge_index);
+                    }
+                }
+                // If neighbor -> current_vertex exists
+                if (has_edge(neighbor, current_vertex)) {
+                    const auto edge_index = get_edge_index(neighbor, current_vertex);
+                    // If edge is breakable throw error
+                    if (get_edge(edge_index).breakable) {
+                        throw std::runtime_error(
+                                "This should never happen, but I found a breakable edge in an unbreakable section");
+                    }
+                    // If edge is not yet in section add it
+                    if (std::find(ret_val.back().begin(), ret_val.back().end(), edge_index) == ret_val.back().end()) {
+                        ret_val.back().emplace_back(edge_index);
+                    }
+                }
+            }
+        }
+    }
+}
+
+void cda_rail::Network::change_vertex_type(int index, cda_rail::VertexType new_type) {
+    /**
+     * Changes the type of a specific vertex.
+     *
+     * @param index: Index of the vertex to change
+     * @param new_type: New type of the vertex
+     */
+
+    if (!has_vertex(index)) {
+        throw std::invalid_argument("Vertex does not exist");
+    }
+    vertices[index].type = new_type;
+}
+
+std::vector<std::pair<int, int>> cda_rail::Network::combine_reverse_edges(const std::vector<int> &edges, bool sort) const {
+    /**
+     * Given a vector of edges, this function combines edges that are the reverse of each other.
+     * If no reverse of an edge exists, the second element will be -1, i.e., negative.
+     * A pair is only added once, i.e., the first index is always smaller than the second for uniqueness.
+     * The order of the edges is not preserved.
+     * It throws an error, if one of the edges does not exist.
+     *
+     * @param edges: Vector of edge indices
+     * @param sort: If true, the pairs are sorted by the neighboring relation. This only works if the edges correspond to a simple path!
+     * @return: Vector of pairs of edge indices
+     */
+
+    // Check if all edges exist
+    for (const auto& edge_index: edges) {
+        if (!has_edge(edge_index)) {
+            throw std::invalid_argument("Edge does not exist");
+        }
+    }
+
+    // Initialize return value
+    std::vector<std::pair<int, int>> ret_val;
+
+    // Iterate over all edges
+    for (const auto& edge_index: edges) {
+        const auto reverse_edge_index = get_reverse_edge_index(edge_index);
+        if (reverse_edge_index < 0 || reverse_edge_index > edge_index) {
+            ret_val.emplace_back(edge_index, reverse_edge_index);
+        }
+    }
+
+    if (sort) {
+        return sort_edge_pairs(ret_val);
+    }
+    return ret_val;
+}
+
+int cda_rail::Network::get_reverse_edge_index(int edge_index) const {
```

Comment:
Sentinel values like -1, are not good practice from a type-safety perspective. Consider using `std::optional` instead.

Reply:
Done